crates/ra_hir/src/code_model.rs

```diff
@@ -1362,6 +1362,98 @@ impl Type {
+            for ty in substs.iter() {
+                walk_type(db, ty, acc);
+            }
+        }
+
+        fn walk_type(db: &dyn HirDatabase, ty: &Ty, acc: &mut Vec<AdtOrTrait>) {
+            match ty.strip_references() {
+                Ty::Apply(ApplicationTy { ctor, parameters, .. }) => {
+                    match ctor {
+                        TypeCtor::Adt(adt_id) => push_new_item(Adt::from(*adt_id).into(), acc),
+                        TypeCtor::AssociatedType(type_alias_id) => {
+                            let trait_id = match type_alias_id.lookup(db.upcast()).container {
+                                AssocContainerId::TraitId(it) => it,
+                                _ => panic!("not an associated type"),
+                            };
+
+                            push_new_item(Trait::from(trait_id).into(), acc);
+                        }
+                        _ => (),
+                    }
+                    // adt params, tuples, etc...
+                    walk_substs(db, parameters, acc);
+                }
+                Ty::Dyn(predicates) => {
+                    push_bounds(db, predicates, acc);
+                }
+                Ty::Placeholder(id) => {
+                    let generic_params = db.generic_params(id.parent);
+                    let param_data = &generic_params.types[id.local_id];
+                    match param_data.provenance {
+                        hir_def::generics::TypeParamProvenance::ArgumentImplTrait => {
+                            let predicates: Vec<_> = db
+                                .generic_predicates_for_param(*id)
+                                .into_iter()
+                                .map(|pred| pred.value.clone())
+                                .collect();
+                            push_bounds(db, &predicates, acc);
+                        }
+                        _ => (),
+                    }
+                }
+                Ty::Opaque(opaque_ty) => {
+                    let bounds = match opaque_ty.opaque_ty_id {
+                        OpaqueTyId::ReturnTypeImplTrait(func, idx) => {
+                            let datas = db
+                                .return_type_impl_traits(func)
+                                .expect("impl trait id without data");
+                            let data = (*datas)
+                                .as_ref()
+                                .map(|rpit| rpit.impl_traits[idx as usize].bounds.clone());
+                            data.clone().subst(&opaque_ty.parameters)
+                        }
+                    };
+                    push_bounds(db, &bounds.value, acc);
+                    walk_substs(db, &opaque_ty.parameters, acc);
+                }
+                _ => (),
+            }
+        }
+
+        let mut res: Vec<AdtOrTrait> = Vec::new(); // not a Set to preserve the order
+        walk_type(db, &self.ty.value, &mut res);
+        res
+    }
```

Comment:
Hm, I am somewhat unhappy that we need to provide a lot of very specific code *at this level of abstraction*. Perhaps we need something more general, like `fn walk_components(&self, cb: impl FnMut(Type))`?

Reply:
But I'd defer to @flodiebold here :)

Reply:
Otherwise, LGTM!

Reply:
>  Perhaps we need something more general...

Yeah, looks like it is possible.

I did not plan to support traits at all, just wanted to get ADTs.
But then noticed that I can support `-> impl Trait`, then `dyn Trait`... and got carried away :)
So I did not think about a visitor\walker at all. I'm not sure that I didn't miss some edge cases to make this code general.

Reply:
Yeah, I was imagining adding a `walk` like method to `Type`. It will require cloning the `Ty` to wrap it into a `Type`, but that should be fine (cloning `Ty` is cheap, and this should not be a hot API anyway). And then we'd need separate methods to get the `dyn Trait` and `impl Trait` bounds.

Reply:
> And then we'd need separate methods to get the dyn Trait and impl Trait bounds.

I think this is a problem, as it is important to preserve the order.
For `impl Trait<Box<dyn AnotherTrait<Struct>>` I'd prefer to see  
`Trait | Box | AnotherTrait | Struct`
and not  
`Box | Struct | Trait | AnotherTrait` or `Trait | AnotherTrait | Box | Struct`

I have no idea how it could be possible if we have separate methods.

Reply:
maybe be we can have `walk_components(&self, cb: impl FnMut(ModuleDef))`? Obviously ModuleDef will be only ModuleDef::Adt or ModuleDef::Trait

Reply:
I mean, the code collecting the components would use those methods while it's walking the type. So the order would still be the same.

Reply:
Oh, I misunderstood you.  Are you talking about the callbacks for each separate entity type? Might be we can group them in a trait? I am trying the following now:

```rust
pub trait TypeComponentVisitor {
    fn visit_type(&mut self, it: &Type);
    fn visit_dyn_trait(&mut self, r: &Trait);
    fn visit_impl_trait(&mut self, r: &Trait);
}
```

```rust
pub fn walk_components(&self, db: &dyn HirDatabase, visitor: &mut impl TypeComponentVisitor)
```

Reply:
Or even better for my needs:

```rust
pub trait TypeComponentVisitor {
    fn visit_adt(&mut self, it: &Adt);
    fn visit_dyn_trait(&mut self, r: &Trait);
    fn visit_impl_trait(&mut self, r: &Trait);
}
```

Reply:
No, what I mean is something like (pseudocode)
```rust
let components = Vec::new();
type.walk(|t| {
    if let Some((adt, _)) = t.as_adt() { components.push(adt); }
    else if let Some(trait) = t.dyn_trait() { components.push(trait); }
    else if let Some(trait) = t.impl_trait() { components.push(trait; }
});
```